File: pc_agent/policies/path_policy.py
"""
NEXUM PC Agent — Path Policy
Blocks access to sensitive system directories.
"""

import os
import re
from pathlib import Path
from typing import Tuple

# ── Blocked patterns ──────────────────────────────────────────────────────────

BLOCKED_PATTERNS = [
    # Unix
    r"^/bin", r"^/sbin", r"^/usr/bin", r"^/usr/sbin",
    r"^/System", r"^/Library/System", r"^/private/etc",
    r"^/etc/shadow", r"^/etc/sudoers", r"^/etc/passwd",
    # Windows
    r"(?i)^C:\\Windows\\System32",
    r"(?i)^C:\\Windows\\SysWOW64",
    r"(?i)^C:\\Windows\\WinSxS",
    # Sensitive config
    r"[/\\]\.ssh[/\\]",
    r"[/\\]\.aws[/\\]",
    r"[/\\]\.gnupg[/\\]",
    # Key files
    r"id_rsa$", r"id_ecdsa$", r"id_ed25519$",
    r"\.pem$", r"\.key$",
    # Crypto wallets
    r"wallet\.dat", r"keystore[/\\]",
]

_compiled = [re.compile(p) for p in BLOCKED_PATTERNS]

# ── Allowed workspace roots ───────────────────────────────────────────────────

HOME = Path.home()

ALLOWED_ROOTS = [
    HOME / "Desktop",
    HOME / "Documents",
    HOME / "Downloads",
    HOME / "nexum-workspace",
    Path("/tmp"),
]


def check_path(file_path: str) -> Tuple[bool, str]:
    """
    Returns (allowed, reason).
    True = access permitted, False = access denied.
    """
    try:
        normalized = os.path.normpath(file_path)
    except Exception:
        return False, "Invalid path"

    # Check for path traversal
    if ".." in file_path:
        return False, "Path traversal attempt detected"

    # Check blocked patterns
    for pattern in _compiled:
        if pattern.search(normalized):
            return False, f"Path is in a protected location: {normalized}"

    return True, "OK"


def is_in_workspace(file_path: str) -> bool:
    """Returns True if path is within allowed workspace roots."""
    try:
        p = Path(file_path).resolve()
        return any(str(p).startswith(str(root)) for root in ALLOWED_ROOTS)
    except Exception:
        return False


def sanitize_path(file_path: str) -> str:
    """Remove null bytes and dangerous characters."""
    return file_path.replace("\0", "").strip()

File: src/agent/policies/path-policy.ts
/**
 * NEXUM Path Policy — file system access control
 * Defines which paths are allowed, blocked, or require confirmation.
 */

import os from 'os';
import path from 'path';

const homeDir = os.homedir();

// ── Allowed workspace roots ────────────────────────────────────────────────────

export const ALLOWED_ROOTS: string[] = [
  path.join(homeDir, 'Desktop'),
  path.join(homeDir, 'Documents'),
  path.join(homeDir, 'Downloads'),
  path.join(homeDir, 'nexum-workspace'),
  '/tmp',
  'C:\\Users\\Public\\Documents',
  'C:\\Nexum',
];

// ── Blocked path patterns ──────────────────────────────────────────────────────

const BLOCKED_PATTERNS: RegExp[] = [
  // Unix system dirs
  /^\/bin/, /^\/sbin/, /^\/usr\/bin/, /^\/usr\/sbin/,
  /^\/System/, /^\/Library\/System/, /^\/private\/etc/,
  /^\/etc\/shadow/, /^\/etc\/sudoers/,
  // Windows system dirs
  /^C:\\Windows\\System32/i, /^C:\\Windows\\SysWOW64/i,
  /^C:\\Windows\\WinSxS/i,
  // Sensitive config dirs
  /[\/\\]\.ssh[\/\\]/, /[\/\\]\.aws[\/\\]/, /[\/\\]\.gnupg[\/\\]/,
  /[\/\\]\.config[\/\\]chromium/i, /[\/\\]\.config[\/\\]google-chrome/i,
  // Key files
  /id_rsa$/, /id_ecdsa$/, /id_ed25519$/, /\.pem$/, /\.key$/,
  // Blockchain wallets
  /wallet\.dat/, /keystore[\/\\]/,
];

// ── Path validators ───────────────────────────────────────────────────────────

export interface PathCheckResult {
  allowed: boolean;
  reason: string;
  normalized: string;
}

export function checkPath(filePath: string): PathCheckResult {
  const normalized = path.normalize(filePath);

  // Check blocked patterns
  for (const pattern of BLOCKED_PATTERNS) {
    if (pattern.test(normalized)) {
      return {
        allowed: false,
        reason: `Path is in a protected location: ${normalized}`,
        normalized,
      };
    }
  }

  // Check for path traversal attempts
  if (normalized.includes('..')) {
    return {
      allowed: false,
      reason: 'Path traversal detected',
      normalized,
    };
  }

  // Check if within allowed roots (warning only — not a hard block for general use)
  const inAllowed = ALLOWED_ROOTS.some(root =>
    normalized.toLowerCase().startsWith(root.toLowerCase())
  );

  return {
    allowed: true,
    reason: inAllowed ? 'Within allowed workspace' : 'Outside default workspace (proceed with caution)',
    normalized,
  };
}

export function isWithinWorkspace(filePath: string): boolean {
  const normalized = path.normalize(filePath);
  return ALLOWED_ROOTS.some(root =>
    normalized.toLowerCase().startsWith(root.toLowerCase())
  );
}

export function sanitizePath(filePath: string): string {
  // Remove null bytes and dangerous chars
  return filePath
    .replace(/\0/g, '')
    .replace(/[<>"|?*]/g, '_')
    .trim();
}
